connectors/kafka/src/main/java/forklift/controller/AcknowledgedRecordHandler.java

```diff
@@ -0,0 +1,121 @@
+ * Acknowledged records are those that have started processing but have not yet been committed to the Kafka Broker.  This class is threadsafe.
+ */
+public class AcknowledgedRecordHandler {
+    private static final Logger log = LoggerFactory.getLogger(AcknowledgedRecordHandler.class);
+    private ConcurrentHashMap<TopicPartition, OffsetAndMetadata> pendingOffsets = new ConcurrentHashMap<>();
+    private Set<TopicPartition> assignment = ConcurrentHashMap.newKeySet();
+    private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    /**
+     * Acknowledges that a record has been received before processing begins.  True is returned if processing should occur else false.
+     * Only records belonging to {@link #addPartitions(java.util.Collection) added partitions} may be processed. Note that this is a
+     * blocking method and a short delay may occur should the available topic paritions be changing.
+     *
+     * @param record the record to acknowledge
+     * @return true if the record has been acknowledged and may be processed, else false
+     * @throws InterruptedException if interrupted
+     */
+    public boolean acknowledgeRecord(ConsumerRecord<?, ?> record) throws InterruptedException {
+        lock.readLock().lock();
+        try {
+            boolean acknowledged;
+            TopicPartition topicPartition = new TopicPartition(record.topic(), record.partition());
+            if (!assignment.contains(topicPartition)) {
+                acknowledged = false;
+            } else {
+                long commitOffset = record.offset() + 1;
+                OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(commitOffset, "Commit From Forklift Server");
+                pendingOffsets.merge(topicPartition, offsetAndMetadata, this::greaterOffset);
+                acknowledged = true;
+            }
+            return acknowledged;
+        } finally {
+            lock.readLock().unlock();
+        }
+    }
+
+    private OffsetAndMetadata greaterOffset(OffsetAndMetadata a, OffsetAndMetadata b) {
+        if (a == null) {
+            return b;
+        } else if (b == null) {
+            return a;
+        }
+        return a.offset() > b.offset() ? a : b;
+    }
+
+    /**
+     * Returns the highest offsets of any acknowledged records.
+     * This is a blocking method as a short delay may occur while any threads which are currently acknowledging records are allowed to
+     * complete and any incoming threads are paused.
+     *
+     * @return a Map of the highest offset data for any acknowledged records
+     * @throws InterruptedException if interrupted
+     */
+    public Map<TopicPartition, OffsetAndMetadata> getAcknowledged() throws InterruptedException {
+        lock.writeLock().lock();
+        try {
+            Map<TopicPartition, OffsetAndMetadata> currentOffsets = new HashMap<>(pendingOffsets.size());
+            currentOffsets.putAll(pendingOffsets);
```

Comment:
Can you just use the copy constructor of `HashMap` e.g.
`return new HashMap<>(pendingOffsets);`